src/wp-includes/class-wp-block-bindings-registry.php

```diff
@@ -0,0 +1,194 @@
+ *
+ * Supports overriding content in blocks by connecting them to different sources.
+ *
+ * @package WordPress
+ * @subpackage Block Bindings
+ * @since 6.5.0
+ */
+
+/**
+ * Core class used for interacting with block binding sources.
+ *
+ *  @since 6.5.0
+ */
+final class WP_Block_Bindings_Registry {
+
+	/**
+	 * Holds the registered block bindings sources, keyed by source identifier.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @var array
+	 */
+	private $sources = array();
+
+	/**
+	 * Container for the main instance of the class.
+	 *
+	 * @since 6.5.0
+	 * @var WP_Block_Bindings_Registry|null
+	 */
+	private static $instance = null;
+
+	/**
+	 * Registers a new block binding source.
+	 *
+	 * Sources are used to override block's original attributes with a value
+	 * coming from the source. Once a source is registered, it can be used by a
+	 * block by setting its `metadata.bindings` attribute to a value that refers
+	 * to the source.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string   $source_name       The name of the source.
+	 * @param array    $source_properties {
+	 *     The array of arguments that are used to register a source.
+	 *
+	 *     @type string   $label              The label of the source.
+	 *     @type callback $get_value_callback A callback executed when the source is processed during block rendering.
+	 *                                        The callback should have the following signature:
+	 *
+	 *                                        `function ($source_args, $block_instance,$attribute_name): mixed`
+	 *                                            - @param array    $source_args    Array containing source arguments
+	 *                                                                              used to look up the override value,
+	 *                                                                              i.e. {"key": "foo"}.
+	 *                                            - @param WP_Block $block_instance The block instance.
+	 *                                            - @param string   $attribute_name The name of an attribute .
+	 *                                        The callback has a mixed return type; it may return a string to override
+	 *                                        the block's original value, null, false to remove an attribute, etc.
+	 * }
+	 * @return array|false Source when the registration was successful, or `false` on failure.
+	 */
+	public function register( $source_name, array $source_properties ) {
+		if ( ! is_string( $source_name ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				__( 'Block binding source name must be a string.' ),
+				'6.5.0'
+			);
+			return false;
+		}
+
+		if ( preg_match( '/[A-Z]+/', $source_name ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				__( 'Block binding source names must not contain uppercase characters.' ),
+				'6.5.0'
+			);
+			return false;
+		}
+
+		$name_matcher = '/^[a-z0-9-]+\/[a-z0-9-]+$/';
```

Comment:
This regex is probably a bit too eager. Only alphanumeric characters and hyphens might be too arbitrarily limited. Underscores, dots or colons are all common in technical names. It also doesn't take into account any non-english characters. 

I think we should only check whether there is a `/` present to indicate a namespace and accept any string before- and after.

On another note, the `_doing_it_wrong()` warnings do not show up in PHP unit tests. I don't know if this is by design or there is another method that could be used here?

Reply:
I copied the same logic that the block types registry uses:

https://github.com/WordPress/wordpress-develop/blob/cabfa6f52cb95930996e34850ccfd9f26569962e/src/wp-includes/class-wp-block-type-registry.php#L64-L81

I thought it would be good to follow the same pattern for consistency.

> On another note, the _doing_it_wrong() warnings do not show up in PHP unit tests. I don't know if this is by design or there is another method that could be used here?

`_doing_it_wrong()` call should fail the test if the following PHPDoc annotation isn't provided:

https://github.com/WordPress/wordpress-develop/blob/cabfa6f52cb95930996e34850ccfd9f26569962e/tests/phpunit/tests/block-bindings/wpBlockBindingsRegistry.php#L72

Reply:
> _doing_it_wrong() call should fail the test if the following PHPDoc annotation isn't provided:

I see, thanks!

> I copied the same logic that the block types registry uses:

OK, I think that's fine in that case! 👍  Let's not forget to mention it in the documentation. I'll update the docstring for `register_block_bindings_source()` & `registry->register()` in https://github.com/WordPress/wordpress-develop/pull/5888 as well.

Reply:
Good point. That would be important to have covered 👍

Reply:
done in f2a611b4f76f2791536d42d0d13b8f49ac13dfd2